crates/ruff_python_formatter/src/expression/expr_attribute.rs

```diff
@@ -15,28 +15,57 @@ impl FormatNodeRule<ExprAttribute> for FormatExprAttribute {
         let ExprAttribute {
             value,
             range: _,
-            attr: _,
+            attr,
             ctx: _,
         } = item;
 
-        let requires_space = matches!(
+        let needs_parentheses = matches!(
             value.as_ref(),
             Expr::Constant(ExprConstant {
                 value: Constant::Int(_) | Constant::Float(_),
                 ..
             })
         );
 
+        if needs_parentheses {
+            value.format().with_options(Parenthesize::Always).fmt(f)?;
+        } else {
+            value.format().fmt(f)?;
+        }
+
+        let comments = f.context().comments().clone();
+
+        if comments.has_trailing_own_line_comments(value.as_ref()) {
+            hard_line_break().fmt(f)?;
+        }
+
+        let dangling_comments = comments.dangling_comments(item);
+
+        let leading_attribute_comments_start =
+            dangling_comments.partition_point(|comment| comment.line_position().is_end_of_line());
+        let (trailing_dot_comments, leading_attribute_comments) =
+            dangling_comments.split_at(leading_attribute_comments_start);
+
         write!(
             f,
             [
-                item.value.format(),
-                requires_space.then_some(space()),
                 text("."),
-                not_yet_implemented_custom_text("NOT_IMPLEMENTED_attr")
+                trailing_comments(trailing_dot_comments),
+                (!leading_attribute_comments.is_empty()).then_some(hard_line_break()),
```

Comment:
oh i didn't realize write would take an option

Reply:
You can also pass an empty tuple. Don't know when that's useful but it's implemented ;)